### PR TITLE
fix(renderer,bcf): guard the four remaining non-finite inputs to camera state

### LINKED ISSUE #2461: camera: ortho zoom and the animator write orthoSize directly, so non-finite bounds can still reach camera state
Follow-up from #2450, which guarded every value that reaches the projection matrix. One gap remains, deliberately left out of that PR because it is a different input class.

`orthoSize` has four writers. #2450 routed the two setters through a shared `usableOrthoSize()` helper and added a read-site backstop in `updateMatrices`, so the projection matrix now stays finite. But the orthographic zoom (`packages/renderer/src/camera-controls.ts:377`) and the camera animator (`camera-animation.ts:105`, `:143`) write `state.orthoSize` **directly**, bypassing the setters. When `frameBounds` / `zoomExtent` are handed non-finite bounds, those writers can still put a non-finite value into camera *state*.

Consequence is narrower than the original defect but not nothing:

- the rendered matrix is fine — the read-site backstop catches it;
- but `getOrthoSize()` can hand back the non-finite value, and that is what gets written into a saved viewpoint, so a bad value can be persisted and restored later.

The real fix is upstream: validate the bounds feeding `frameBounds` / `zoomExtent`. That is a wider blast radius than a viewpoint file — these bounds come from geometry, so the fix has to decide what a non-finite model AABB means (reject, clamp, or fall back to the previous bounds) and that decision affects fit, orbit-anchor and section range too, not just ortho zoom.

Worth checking whether non-finite bounds are actually reachable from a malformed model before designing the guard — #2458 is a cautionary example of a plausible-looking defect that had already been fixed. Reachability first, then the fix.

### LINKED ISSUE #2466: BCF viewpoint restore feeds a raw getDistance() back in, so a malformed viewpoint cannot be recovered from
Found while adversarially reviewing #2463, and traced end to end rather than inferred.

#2463 guards every navigation gesture inside `@ifc-lite/renderer` against a non-finite pose distance, and deliberately leaves `getDistance()` itself raw so a broken pose stays visible to callers. That is the right call for the package, but the app layer does not guard, and it cannot: `isUsableDistance` is package-internal, re-exported nowhere.

The consequence is reachable today, and it is the recovery path specifically:

- `packages/bcf/src/reader.ts` `parsePoint` (~L818-841) extracts `<X>/<Y>/<Z>` with bare `parseFloat` and no finiteness check. `parseFloat("NaN")`, arbitrary garbage, or `1e999` (which yields `Infinity`) all pass through.
- Those flow via `extractViewpointState` → `perspectiveToCamera` → `applyCameraState` (`apps/viewer/src/hooks/useBCF.ts` ~L126-127) into `setPosition`/`setTarget`, which store them verbatim.
- Once the pose is broken, `useBCF.ts` ~L453 and ~L479 feed a raw `getDistance()` into `extractViewpointState` when **restoring** a viewpoint: `target = cameraViewPoint + direction * NaN`.

So restoring a known-good viewpoint — the one action a user would naturally reach for after a bad one wrecks the camera — silently fails to repair it. Only a fit, a preset view or home recovers. Post-#2463 the gestures no longer make it worse, but nothing makes it better.

Two candidate fixes, and the choice is a real design decision rather than an oversight:

1. **Validate at the parser.** Reject or clamp non-finite coordinates in `parsePoint`, so a malformed viewpoint never reaches the camera. Narrow and catches the problem at the boundary, but changes BCF read behaviour and needs a decision about what a malformed viewpoint should do (skip the viewpoint, drop the camera, fall back to bounds).
2. **Export the predicate.** Make `isUsableDistance` public from `@ifc-lite/renderer` so app-layer consumers can guard. Wider surface, and pushes the burden onto every caller — which is what the current gap already demonstrates does not happen by default.

Worth pinning whichever is chosen with a test that starts from a genuinely malformed BCF file, not a hand-poked camera.

Note for whoever picks this up: `Infinity` is a real route here (`1e999`), and it is currently **untested** everywhere — every existing gesture test uses `NaN`. A guard written as `!Number.isNaN(...)` instead of `Number.isFinite(...)` would pass the whole suite and still be wrong.

### LINKED ISSUE #2467: camera-projection: unprojectToRay returns a NaN ray origin for a non-finite camera.up, and picking reads that as a miss
Found while fixing the `up`-vector gap on #2463, and deliberately left out of that PR because it is outside its thesis (a gesture must not spread a malformed pose — this returns a value rather than writing state).

`unprojectToRay`'s orthographic branch (`packages/renderer/src/camera-projection.ts` ~L91) returns `origin = (NaN, NaN, NaN)` when `camera.up` is non-finite — for **both** NaN and Infinity, unlike the gesture paths where the two behave differently.

Why it matters more than a wrong number: this function feeds **picking and measurement**, which live outside the renderer. A NaN ray origin does not throw. Consumers that test a hit with a comparison (`t < tMin`, `dist < tolerance`) will find every comparison false and silently report "no hit", which is indistinguishable from clicking empty space. That is the same shape as the guards this recent work has been unpicking: a comparison is not a finiteness check.

Reachable by the same route proven for #2463: `parseFloat` at `packages/bcf/src/reader.ts:831-833` turns `1e999` into `Infinity`, `perspectiveToCamera` passes it through, and `Viewport.tsx:823` restores viewpoints via `animateToWithUp`, which writes `camera.up` verbatim without going through `setUp`.

The fix needs a decision this issue should not pre-empt: what should an unprojection return when the camera basis is unusable? Returning `null` is honest but pushes a new branch onto every caller; falling back to a world-axis `up` produces a plausible ray that is quietly wrong, which is worse for measurement. Whoever picks this up should enumerate the callers first and let that choose.

Note when testing: use an **off-centre** cursor. At the exact canvas centre both NDC terms are zero, so every `up` produces the same answer and the bug is invisible — that blinded the first probe on #2463.

Related: #2466 (unguarded app-layer `getDistance()` consumers), #2461 (non-finite bounds reaching `orthoSize`).

### LINKED ISSUE #2473: camera-controls validates the pose but not the gesture arguments (deltas, cursor coords, canvas size)
Raised by a review pass over #2463 and deliberately left out of it: that PR's thesis is that a gesture must not spread a malformed **pose**, and this is a different input class — the gesture **arguments**.

`camera-controls.ts` validates the pose (position, target, and now `up`) before acting, but takes its arguments on trust:

- `zoom(delta, mouseX, mouseY, canvasWidth, canvasHeight)` — a non-finite `delta`, non-finite cursor coordinates, or a zero/negative/non-finite canvas dimension all reach the arithmetic. The cursor-anchored branch divides by the canvas dimensions, so a zero width yields a non-finite NDC and the anchor writes it into the target.
- `orbit`, `pan`, `moveFirstPerson` — non-finite deltas propagate into the pose the same way.

The consequence is the same failure #2463 exists to prevent, reached from the other side: a finite pose is destroyed by a bad argument rather than a bad pose destroying a good gesture.

Reachability is the open question and should be settled before designing the guard, not after. Arguments come from event handlers rather than files, so the route is different from the BCF one that made #2463's case: a resize race that reports a zero-height canvas, a synthetic or replayed event, a trackpad delta from an unusual driver, or a caller of the public `@ifc-lite/renderer` API passing whatever it likes. Establish which of those actually occur before hardening all four entry points.

Note the shape that has bitten repeatedly in this area: a magnitude comparison is not a finiteness check (`dist < eps` is false for NaN), and a lower-bound floor rejects NaN while admitting Infinity (`len > 1e-10` is true for Infinity, and `Infinity * 0` is NaN). Whatever guard this gets should use `Number.isFinite`, and the canvas dimensions need `> 0` as well.

`isUsableDistance` and `isUsableUp` already exist in that module and are the natural precedent for an argument-level equivalent.

### PR DESCRIPTION
**Pull request by louistrue**
Fixes #2461
Fixes #2466
Fixes #2467
Fixes #2473

The four remaining members of the family whose earlier members shipped in #2441, #2450 and #2463: non-finite values reaching camera state through paths a magnitude comparison does not guard. They are one PR because they share one predicate set, now in `packages/renderer/src/camera-guards.ts`, and because two of them intersect — a non-finite zoom delta (#2473) is one of the ways a non-finite `orthoSize` gets written (#2461).

Reachability was established before each guard was designed. Where a route turned out not to exist, that is said plainly below rather than papered over.

## #2473 — gesture arguments

**Reachability: the published API only.** No in-app event handler can produce a non-finite argument. Wheel and pointer deltas are browser-guaranteed finite and nothing in the app dispatches a synthetic one; the SpaceMouse path clamps its axes four separate times (`parser.ts` `clampAxis`, the `deltaMs > 0` gate, the `(…) || 0` wrappers, and the sensitivity clamp on both load and set); the pinch delta is a subtraction, not a division, so the classic `newDist / oldDist` divide-by-zero does not exist here; first-person and arrow-pan pass compile-time literals; and every canvas writer floors the drawing buffer (`Math.max(1, …)` with a 64px width floor) *before* `new Renderer(canvas)`, with the input hooks gated on `isInitialized`. A fresh `<canvas>` is 300x150, not 0x0.

What *is* real is the published surface. `docs/guide/rendering.md`, `docs/guide/quickstart.md` and the `create-ifc-lite` React template all document these methods as the input contract and wire raw `e.clientX` deltas and `e.deltaY` straight into them, so an embedder's own DPR scaling, gamepad axis or custom pinch ratio is an unvalidated path into arithmetic that writes `position`, `target` and `orthoSize`. `Renderer.resize` is the same class: `canvas.width` is an IDL `unsigned long`, so `resize(NaN, NaN)` coerces to **0**, and no pick guard in the package checks the drawing buffer — they all check the bounding rect. Both in-repo callers already floor their own values; an external ResizeObserver would not.

**Measured on the pre-guard code**, all from a finite pose: `orbit(NaN, 0)`, `orbit(Infinity, 0)`, `pan(NaN, 0)`, `pan(Infinity, 0)`, `moveFirstPerson(NaN, 0, 0)`, `moveFirstPerson(Infinity, 0, 0)` and `zoom(NaN)` each returned a non-finite position. `zoom(Infinity)` did **not** — its clamp `Math.min(|d| * s, MAX_ZOOM_DELTA)` absorbs Infinity but not NaN. That asymmetry is why both flavours are tested everywhere.

**Design.** Reject and leave the pose untouched, at the package boundary rather than at each of the documented entry points. Three details worth the review:

- The `orbit` guard sits *before* the `camera.up = worldY` reset, so a rejected gesture changes nothing at all rather than half-applying a side effect.
- The cursor anchor now needs `Number.isFinite(w) && w > 0`, not truthiness. Truthiness already rejected `0` and `NaN` (both falsy) — which is why a collapsed canvas never reached the division — but accepted a **negative** width, which mirrors the anchor, and `Infinity`, which pins it to a screen edge. The cursor coordinates were never checked at all. An unusable extent now degrades to the un-anchored zoom; the zoom still happens.
- The three `add*Velocity` accumulators are guarded in their own right. They accumulate in place and the inertia loop spends a velocity only while `Math.abs(v) > minVelocity` — false for NaN — so one poisoned argument would latch orbit, pan or zoom inertia dead for the session, never applied and never decaying. That is the `walkVelocity` latch from #2441 reached from the argument side.

One in-repo site is fixed too: `Viewport.tsx`'s `unprojectToFloor` was the only client-to-drawing-buffer scaler in the repo without a `rect.width > 0` guard — its five siblings (`selectionHandlers`, `picking-manager`, `raycast-engine`, `CesiumPlacementEditor`, `projectScreen`) all have one.

## #2461 — non-finite bounds reaching `orthoSize` and the pose

**Reachability: yes, from geometry, and it does not even need a NaN.** Every AABB accumulator in the renderer seeds `min = +Infinity, max = -Infinity` and narrows on a bare comparison, which is false for a non-finite vertex. `Scene.addMeshData` validates nothing — no length check, no finiteness check — so a mesh piece with zero-length positions makes `Scene.getEntityBoundingBox` return, **and cache**, the inverted-empty sentinel `{min: +Inf, max: -Inf}` as if it were a real box. `model-bounds-tracker.test.ts` already asserts exactly that value as the tracker's output for "a mesh with no finite vertex". Four call sites hand such a box to the camera with only a null check in between: BCF zoom-to-topic, instanced frame-selection, clash framing and the point-cloud fit. (The streaming auto-fit is *not* exposed — `robustFitBoundsAccumulator` and `computeBounds` both filter and return null.)

Upstream of that, non-finite vertices are a documented reality in this repo, not a hypothesis: #1645 exists because `IfcDirection((0,0,0))` produced a NaN placement matrix, and its own comment says so in the present tense. One caller still routes around that fix — `profile_extractor.rs` normalizes before handing the axis to the hardened builder — and `try_normalize`'s `n <= min_norm` test is false for NaN, so the guard cannot catch it. Separately, `Mesh.positions` is `Vec<f32>`, so a legal finite `1.E300` coordinate saturates to `f32::INFINITY` on the cast with no magnitude check anywhere.

**Measured on the pre-guard code**, orthographic, from a healthy camera: `zoomExtent`/`frameBounds` with a NaN corner gave `orthoSize = NaN` and position `(NaN, NaN, NaN)`; with an infinite span, `orthoSize = Infinity`. `fitToBounds` and `setPresetView` destroyed the pose the same way. And `zoom(NaN)` in orthographic mode left `getOrthoSize()` returning NaN — the issue's core claim, confirmed.

**Design decision — reject, not clamp, not fall back.** There is no meaningful clamp for an infinite or inverted extent (clamp it to *what*?), and these entry points are stateless, so there are no "previous bounds" available at that layer. Keeping the pose the user already had is strictly better than moving to one nothing can recover from, and it is the policy `setAspect`, `setFOV` and `setOrthoSize` already settled on in #2441/#2450. The guard rejects non-finite corners **and** an inverted box (`max < min` on any axis), because that is the empty-AABB sentinel leaking out; a *degenerate* box (`max === min` — a flat wall, a single point, a plan-only model) is explicitly still framed, and there is an anti-mutation for it.

The issue notes the decision affects more than ortho zoom, and it does: the same predicate now guards fit, the orbit-anchor path via `framePoint`, and the ViewCube presets. `setPresetView` checks the bounds it *derived* as well as the ones it was given, since `getCurrentBounds()` reads the live pose. `fitBoundsAdaptive` cannot no-op — callers read its return value — so it reports the camera's current pose as a policy, which callers apply as a no-op and read `.kind` from as before.

The two writers that bypassed `setOrthoSize` now go through the same clamp. That catches a case the delta guard does not: a legitimately huge but finite half-height scaled by one more zoom-out notch overflows to `Infinity`, and `Math.max(0.01, Infinity)` forwards it.

**Deliberately out of scope:** `Scene.getEntityBoundingBox` still caches the sentinel. Fixing the Scene cache changes what raycasting, section ranges and clash all see, which is a different blast radius and belongs in its own issue.

## #2467 — `unprojectToRay` returned a NaN ray origin

**Reachability: confirmed, and for both flavours.** Orthographic branch, off-centre cursor, `up = (NaN, 1, 0)` and `up = (Infinity, 1, 0)` both returned `origin = (NaN, NaN, NaN)` while the view-projection matrix stayed perfectly finite. `MathUtils.normalize` floors on `len < 1e-10`, which is false for NaN, so it divided by NaN instead of falling back — and Infinity got through the other way. Reachable by the route #2463 proved: `1e999` in a BCF `CameraUpVector`, through `perspectiveToCamera`, restored by `animateToWithUp`, which writes `camera.up` without going near `setUp`. Also confirmed: a non-finite camera *position* poisoned the origin in both branches, and a zero-extent canvas produced `±Infinity` NDC.

**Design — callers enumerated first, as the issue asked.** Eight call sites: three inside the package (`picking-manager`, `raycast-engine` x2) and five in the app (`selectionHandlers`, `useMouseControls`, `useTouchControls`, `CesiumPlacementEditor`, `Viewport`). `null` would put a branch on all eight, including consumers of a published API, and would fail a pick on a frame the user can see and interact with.

What settles it is not that count, though — it is that **`MathUtils.lookAt` already substitutes a deterministic basis** when `up` carries no usable orientation (world Y, or -Z for a vertical view). So there *is* a well-defined frame on screen even for a malformed `up`, and the honest thing for a ray is to belong to that frame. A world-axis fallback invented locally would be "plausible but quietly wrong"; deriving the ray from *the same basis the view matrix is built from* is neither. `MathUtils.viewBasis` is extracted so `lookAt` and `unprojectToRay` cannot drift apart, and the test asserts the ray origin is reconstructible from it exactly. Same reasoning gives the origin the scrubbed eye, so a ray never belongs to a pose that was never rendered, and gives a zero-extent viewport a centred cursor rather than an infinite one.

Every probe is off-centre by construction, with a test pinning that a centred cursor is the one case where the basis legitimately does not matter — the vacuity trap that blinded the first probe on #2463.

## #2466 — a viewpoint restore could not repair a broken pose

**Two halves, and they are different failures.** Both fixes are one-place fixes; neither is "guard the six consumers", which is the arrangement that produced the gap.

**Prevention, at the parser.** `parsePoint` used a bare `parseFloat`, which has no out-of-band failure value: `"NaN"` parses to `NaN` and the well-formed literal `"1e999"` parses to `Infinity`. This is the issue's option 1, and it turns out to cost *nothing*: all five `parsePoint` call sites already treat `undefined` as "drop the thing I was reading" for a missing element, so reporting an unusable number the same way adds no new branch anywhere. A malformed camera is dropped and the rest of the viewpoint — selection, visibility, clipping, snapshot — still applies, which is the right answer for a viewpoint that is more than its camera. `FieldOfView` and `ViewToWorldScale` get the same treatment; the latter becomes the orthographic half-height that `getOrthoSize()` writes into a saved viewpoint.

Option 2 (exporting `isUsableDistance`) is rejected, for the reason the issue itself gives: it pushes the burden onto every caller, which demonstrably does not happen by default.

**Recovery, at the sink.** Prevention alone does not fix the reported defect, which is specifically that restoring a *known-good* viewpoint could not repair a camera already broken by some other route — and #2473 shows such routes exist. Rather than guard the two `useBCF` call sites, the guard goes where every restore path funnels: `perspectiveToCamera`, `orthogonalToCamera` and `computeMarkerPositions` all take the viewer's raw `camera.getDistance()` and multiply a direction by it. They now fall back to their own documented defaults when it is not usable. Zero is rejected alongside the non-finite values, since a zero reference distance collapses the target onto the eye. `computeMarkerPositions` is included because its output drives `Camera.framePoint` — it is a pose writer too, and a destructuring default only fires for `undefined`, never for the `NaN` a broken pose actually supplies.

`Camera.getDistance()` stays raw, as #2463 intended.

The parser tests start from the real buildingSMART fixtures with exactly one number edited and the archive re-zipped, per the issue's request — not from a hand-poked camera.

## Verification

`npx turbo typecheck --force` — **36 successful, 36 total; Cached: 0 cached, 36 total.**
`npx turbo test --force` — **86 successful, 86 total; Cached: 0 cached, 86 total.**

Renderer 649 leaf tests (625 on main + 24 added), 0 fail. BCF 171 (150 + 21), 0 fail. Viewer 3388, 0 fail. `pnpm lint` and all three `scripts/docs/check-*` gates green.

**Mutation evidence.** Every new guard was reverted individually, the suite re-run, and the specific test confirmed to fail on leaf counts; each mutation was proved to have applied (`git diff` non-empty) and each restore proved to have landed (`git status --porcelain` empty). Baselines: 20 / 9 / 9 / 21 leaves.

| Mutation | Result |
|---|---|
| gesture argument guards removed | 20 tests, **2 fail** — the delta test and the inertia-latch test |
| velocity accumulator guards removed | 20 tests, **1 fail** — the inertia-latch test |
| cursor anchor back to truthiness | 20 tests, **2 fail** — cursor coords and canvas extent |
| `unprojectToRay` back to its own basis | 9 tests, **4 fail** |
| NDC extent guard removed | 9 tests, **1 fail** |
| all `isUsableBounds` guards removed | 9 tests, **2 fail** |
| `framePoint` + `frameBounds` guards removed | 9 tests, **1 fail** |
| `orthoSize` writers back to `Math.max` | 9 tests, **1 fail** — the overflow case |
| zoom-delta **and** `orthoSize` clamp both removed | 9 tests, **2 fail** |
| `parseFiniteFloat` back to bare `parseFloat` | 21 tests, **14 fail** |
| `usableTargetDistance` guard removed | 21 tests, **4 fail** |

**Anti-mutations**, proving an over-broad guard would be caught:

| Over-broad guard | Result |
|---|---|
| cursor anchor also requires `0 < cursor < extent` | 20 tests, **1 fail** — the edge-cursor control |
| NDC extent floor raised to `> 1024` | 9 tests, **4 fail** |
| bounds guard uses `max > min` (rejects a flat box) | 9 tests, **1 fail** — the degenerate-box control |
| parser also rejects `\|v\| >= 1e6` and `0` | 21 tests, **2 fail** |
| distance fallback always fires | 21 tests, **1 fail** |

One of those anti-mutations initially **passed**, which is the reason for commit `2cc11353`. The edge-cursor control asserted "the target moved", and perspective zoom dollies the target forward on its own — so it moved whether the anchor was honoured or dropped. It now compares against the un-anchored zoom, and catches the over-broad guard.

## Notes

`camera.ts`, `camera-animation.ts` and `camera-controls.ts` remain over the ~400-line rule; that is pre-existing and tracked as #2460, and nothing here is split for it. `camera-guards.ts` is new because four modules now share these predicates, and it takes ~60 lines out of `camera.ts` and `camera-controls.ts` on the way.

Changeset: `minor` for both packages. Additive on `@ifc-lite/renderer` (`MathUtils.viewBasis`, which is exported), and both packages change observable behaviour for malformed input — a BCF file that previously yielded a NaN camera now yields no camera, which a consumer could notice.

## Worth a human eye

Nothing here needs a GPU — every path is CPU camera math and file parsing, and all of it is covered headlessly. If you want one browser check, the highest-value one is a **cursor-anchored wheel zoom at each of the four viewport corners** on localhost, confirming the anchor still tracks the cursor: the edge cursors (`0` and the exact canvas dimension) are the values the over-broad-guard anti-mutation was written to protect, and they are the only place a mistake here would be visible rather than merely provable.

### CODE REVIEW

File: packages/renderer/src/math.ts [hunk 1]
```diff
@@ -116,18 +116,43 @@ export class MathUtils {
...
+     * `eye` is returned alongside the axes because it is scrubbed here too,
+     * and a ray origin must come from the same scrubbed eye the view matrix's
+     * translation row does.
+     *
+     * All four vectors are always finite. `right`/`up` are unit length;
+     * `forward` points from eye towards target (the negation of `lookAt`'s
+     * third row, which is the camera's backward axis).
      */
-    static lookAt(eye: Vec3, target: Vec3, up: Vec3): Mat4 {
+    static viewBasis(eye: Vec3, target: Vec3, up: Vec3): {
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/renderer/src/math.ts]: Taken the first option — `viewBasis` is now a module-level export from `math.ts` rather than a static on `MathUtils`, and `index.ts` does not re-export it, so it stays internal. Both production callers (`lookAt` and `unprojectToRay`) are inside this package, so there was nothing for a consumer to use and no surface worth documenting.

Worth recording for the next time this comes up: `check-api-surface.mjs` reports **4064 exports either way**. It compares names and kinds, so a new static method on an already-exported class does not move the snapshot — the gate would not have caught this, which is why the review did.

Renderer suite 649/649/0, typecheck green, API surface unchanged.
